File: game-data/src/main/kotlin/io/github/lishangbu/gamedata/entity/GameCreature.kt
package io.github.lishangbu.gamedata.entity

import io.github.lishangbu.common.persistence.jimmer.CosIdLongUserIdGenerator
import org.babyfish.jimmer.jackson.JsonConverter
import org.babyfish.jimmer.jackson.LongToStringConverter
import org.babyfish.jimmer.sql.Entity
import org.babyfish.jimmer.sql.GeneratedValue
import org.babyfish.jimmer.sql.Id
import org.babyfish.jimmer.sql.Table

/**
 * 精灵资料持久化实体。
 *
 * 对应 `game_creature` 表，仅承载一条精灵资料记录及其标量属性。
 * 关联项保留为独立 ID，避免把其他功能的实体聚合进当前资料对象；新增记录的主键统一由 CosId 生成。
 */
@Entity
@Table(name = "game_creature")
interface GameCreature {
	/**
	 * 精灵资料记录的唯一标识。
	 *
	 * 新增时由 [CosIdLongUserIdGenerator] 生成；对外 JSON 通过 [LongToStringConverter] 输出为字符串，
	 * 避免 JavaScript 数值精度丢失，同时在 Kotlin 和数据库中保持 Long 类型。
	 */
	@Id
	@GeneratedValue(generatorType = CosIdLongUserIdGenerator::class)
	@JsonConverter(LongToStringConverter::class)
	val id: Long

	/**
	 * 精灵资料的编码。
	 *
	 * 对应数据库 `code` 列，写入时必须提供，最大长度为 80 个字符。
	 */
	val code: String

	/**
	 * 精灵资料的名称。
	 *
	 * 对应数据库 `name` 列，写入时必须提供，最大长度为 120 个字符。
	 */
	val name: String

	/**
	 * 精灵资料引用的种类 ID。
	 *
	 * 对应数据库 `species_id` 列，写入时必须提供；关联标识以 Long 标量保存，不在该实体中聚合关联对象。
	 */
	val speciesId: Long

	/**
	 * 精灵资料继承缺失数据时引用的来源精灵 ID。
	 *
	 * 对应数据库 `inherits_from_creature_id` 列，允许为空；数据库约束要求来源精灵属于同一种类且不能自引用。
	 */
	val inheritsFromCreatureId: Long?

	/**
	 * 精灵资料的高度。
	 *
	 * 对应数据库 `height` 列，允许为空。
	 */
	val height: Int?

	/**
	 * 精灵资料的重量。
	 *
	 * 对应数据库 `weight` 列，允许为空。
	 */
	val weight: Int?

	/**
	 * 精灵资料的基础经验。
	 *
	 * 对应数据库 `base_experience` 列，允许为空。
	 */
	val baseExperience: Int?

	/**
	 * 精灵资料的排序。
	 *
	 * 对应数据库 `sort_order` 列，允许为空。
	 */
	val sortOrder: Int?

	/**
	 * 用于标识精灵资料的“默认形态”状态。
	 *
	 * 对应数据库 `default_form` 列，允许为空。
	 */
	val defaultForm: Boolean?

	/**
	 * 用于标识精灵资料的“启用”状态。
	 *
	 * 对应数据库 `enabled` 列，允许为空。
	 */
	val enabled: Boolean?
}

File: ent/schema/game_creature.go
package schema

import (
	"entgo.io/ent"
	"entgo.io/ent/dialect/entsql"
	"entgo.io/ent/schema"
	"entgo.io/ent/schema/field"
	"github.com/lishangbu/avalon/internal/platform/snowflake"
)

// GameCreature 定义 game_creature 表的持久化结构。
type GameCreature struct {
	ent.Schema
}

// Fields 返回 game_creature 表全部字段及其数据库约束。
func (GameCreature) Fields() []ent.Field {
	return []ent.Field{
		field.Int64("id").GoType(snowflake.ID(0)).Positive().Annotations(nonIncrementalSnowflakeIdentifier()).Comment("该资料记录的稳定 Identifier。"),
		field.String("code").MaxLen(64).Comment("该资料记录的全局唯一稳定编码。"),
		field.String("name").MaxLen(120).Comment("该资料记录的简体中文显示名称。"),
		field.Int64("species_id").GoType(snowflake.ID(0)).Positive().Comment("该资料记录的 species id 业务属性。"),
		field.Int64("inherits_from_creature_id").GoType(snowflake.ID(0)).Positive().Optional().Nillable().Comment("该资料记录的 inherits from creature id 业务属性。"),
		field.Int32("height").Optional().Nillable().Comment("该资料记录的 height 业务属性。"),
		field.Int32("weight").Optional().Nillable().Comment("该资料记录的 weight 业务属性。"),
		field.Int32("base_experience").Optional().Nillable().Comment("该资料记录的 base experience 业务属性。"),
		field.Int32("capture_rate").Optional().Nillable().Comment("该资料记录的 capture rate 业务属性。"),
		field.Int32("hatch_cycles").Optional().Nillable().Comment("该资料记录的 hatch cycles 业务属性。"),
		field.Int16("male_eighths").Comment("该资料记录的 male eighths 业务属性。"),
		field.Int16("female_eighths").Comment("该资料记录的 female eighths 业务属性。"),
		field.Bool("default_form").Comment("该资料记录的 default form 业务属性。"),
		field.Bool("enabled").Comment("是否允许新的业务数据引用该实时资料。"),
		field.Int64("version").Annotations(entsql.DefaultExpr("1")).Comment("该资料记录的乐观并发控制版本。"),
		field.Time("created_at").Annotations(entsql.DefaultExpr("CURRENT_TIMESTAMP")).Comment("该资料记录的创建时间。"),
		field.Time("updated_at").Annotations(entsql.DefaultExpr("CURRENT_TIMESTAMP")).Comment("该资料记录的最后更新时间。"),
	}
}

// Annotations 固定 game_creature 的表名、注释、复合主键和检查约束。
func (GameCreature) Annotations() []schema.Annotation {
	return []schema.Annotation{
		schema.Comment("实时游戏资料表。"),
		entsql.WithComments(true),
		entsql.Annotation{Table: "game_creature", Checks: map[string]string{
			"game_creature_capture_rate_check": "capture_rate IS NULL OR capture_rate >= 0 AND capture_rate <= 255",
			"game_creature_code_check":         "code::text ~ '^[a-z][a-z0-9-]{1,63}$'::text",
			"game_creature_gender_ratio_check": "(male_eighths = 0 AND female_eighths = 0 OR (male_eighths + female_eighths) = 8) AND male_eighths >= 0 AND male_eighths <= 8 AND female_eighths >= 0 AND female_eighths <= 8",
			"game_creature_hatch_cycles_check": "hatch_cycles IS NULL OR hatch_cycles >= 0",
			"game_creature_inheritance_check":  "inherits_from_creature_id IS NULL OR inherits_from_creature_id <> id",
			"game_creature_name_check":         "char_length(name::text) >= 1 AND char_length(name::text) <= 120 AND name::text = btrim(name::text)",
			"game_creature_size_check":         "(height IS NULL OR height >= 0) AND (weight IS NULL OR weight >= 0) AND (base_experience IS NULL OR base_experience >= 0)",
			"game_creature_version_check":      "version > 0",
		}},
	}
}
